akka-persistence-typed/src/test/java/jdocs/akka/persistence/typed/AccountExample.java

```diff
@@ -0,0 +1,134 @@
+    public final double balance;
+
+    OpenedAccount(double balance) {
+      this.balance = balance;
+    }
+  }
+  public static class ClosedAccount implements Account {}
+
+  public static Behavior<AccountCommand> behavior(String accountNumber) {
+    return Behaviors.setup(context -> new AccountExample(context, accountNumber));
+  }
+
+  public AccountExample(ActorContext<AccountCommand> context, String accountNumber) {
+    super(accountNumber);
+  }
+
+  @Override
+  public Account emptyState() {
+    return new EmptyAccount();
+  }
+
+  private CommandHandlerBuilder<AccountCommand, AccountEvent, EmptyAccount, Account> initialHandler() {
+    return commandHandlerBuilder(EmptyAccount.class)
+      .matchCommand(CreateAccount.class, (__, cmd) -> Effect().persist(new AccountCreated()));
+  }
+
+  private CommandHandlerBuilder<AccountCommand, AccountEvent, OpenedAccount, Account> openedAccountHandler() {
+    return commandHandlerBuilder(OpenedAccount.class)
+      .matchCommand(Deposit.class, (__, cmd) -> Effect().persist(new Deposited(cmd.amount)))
+      .matchCommand(Withdraw.class, (acc, cmd) -> {
+        if ((acc.balance - cmd.amount) < 0.0) {
+          return Effect().unhandled(); // TODO replies are missing in this example
+        } else {
+          return Effect().persist(new Withdrawn(cmd.amount))
+            .andThen(acc2 -> { // FIXME in scaladsl it's named thenRun, change javadsl also?
+              // we know this cast is safe, but somewhat ugly
+              OpenedAccount openAccount = (OpenedAccount) acc2;
+              // do some side-effect using balance
+              System.out.println(openAccount.balance);
+            });
+        }
+      })
+      .matchCommand(CloseAccount.class, (acc, cmd) -> {
+        if (acc.balance == 0.0)
+          return Effect().persist(new AccountClosed());
+        else
+          return Effect().unhandled();
+        });
+  }
+
+  private CommandHandlerBuilder<AccountCommand, AccountEvent, ClosedAccount, Account> closedHandler() {
+    return commandHandlerBuilder(ClosedAccount.class)
+        .matchCommand(AccountCommand.class, (__, ___) -> Effect().unhandled());
+  }
+
+  @Override
+  public CommandHandler<AccountCommand, AccountEvent, Account> commandHandler() {
+    return initialHandler()
+      .orElse(openedAccountHandler())
+      .orElse(closedHandler())
+      .build();
+  }
+
+  @Override
+  public EventHandler<Account, AccountEvent> eventHandler() {
+    return eventHandlerBuilder()
+      .matchEvent(AccountCreated.class, EmptyAccount.class, (__, ___) ->
+          new OpenedAccount(0.0))
+      .matchEvent(Deposited.class, OpenedAccount.class, (acc, cmd) ->
+          new OpenedAccount(acc.balance + cmd.amount))
+      .matchEvent(Withdrawn.class, OpenedAccount.class, (acc, cmd) ->
+          new OpenedAccount(acc.balance - cmd.amount))
+      .matchEvent(AccountClosed.class, OpenedAccount.class, (acc, cmd) ->
+          new ClosedAccount())
+    .build();
```

Comment:
A nice convention when using ADT style state like in this example is to have this in the state classes, like in https://github.com/akka/akka/pull/25503/files#diff-e6c5cde1abab80a30f23b6d59e7bb0d0R29

To support that we would need another EventHandlerBuilder that is made for that purpose.

Reply:
created issue https://github.com/akka/akka/issues/25672